docs/proposals/20230718-inplace-workload-vertical-scaling.md

```diff
@@ -0,0 +1,813 @@
++                       klog.ErrorS(err, "Failed to update pod cache", "pod", klog.KObj(pod))
++                       return false, err
++               }
++       }
++
+        return false, nil
+ }
+```
+
+The corresponding changes to validation are not expanded here.
+
+### Goals
+
+- Users only need to modify template in workload to achieve vertical scaling for all managed Pods
+- Simple and clear configuration strategy, support grayscale capabilities
+- Provides the rollback function, and attempts to return to the previous state when the capacity expansion and shrinkage fails. If the rollback still fails, no further action is required
+
+### Non-Goals/Future Work
+
+- Implement preemption ensures that all Pods can successfully expand and shrink
+
+## Proposal
+
+### User Stories
+
+Users can start vertical scaling of all managed Pods by modifying the resource field in workload.
+
+### Implementation Details/Notes/Constraints
+
+#### Design Objectives
+
+- Use vscale as a standalone operation rather than integrated in update. For example, for cloneset, three operations will be improved to scale, vscale, and update.
+- Adding the vscaleStrategy field to workload requiring vscale operations allows users to configure strategies to handle different scenarios.
+
+#### CloneSet
+
+##### Extend Definition
+
+We want users to write in PodSpec whether or not each container needs to be restarted.
+
+```go
+// CloneSetVScaleStrategy defines strategies for pods vscale.
+type CloneSetVScaleStrategy struct {
+	MaxProposed *intstr.IntOrString `json:"maxProposed,omitempty"`
+	Paused   bool  `json:"paused,omitempty"`
+}
+```
+
+##### Enhance Expectation
+
+Create a new VScaleExpectations for use by different controllers.
+
+```go
+// VScaleExpectations is an interface that allows users to set and wait on expectations of pods scale.
+type VScaleExpectations interface {
+	ExpectVScale(controllerKey string, action ScaleAction, name string)
+	ObserveVScale(controllerKey string, action ScaleAction, name string)
+	SatisfiedExpectations(controllerKey string) (bool, time.Duration, map[ScaleAction][]string)
+	DeleteExpectations(controllerKey string)
+	GetExpectations(controllerKey string) map[ScaleAction]sets.String
+}
+```
+
+##### Enhance calculateDiffsWithExpectation
+
+Vscale-related parameters need to be added to expectationDiffs
+
+```go
+type expectationDiffs struct {
+	...
+	// vscale identifies whether resources have changed
+	vscale bool
+}
+```
+
+At the same time need to modify the `calculateDiffsWithExpectation` function, when the vscale closed scale behavior and the behavior of the update.
+
+##### Flow Chart
+
+The general workload processing logic is shown in the following figure.
+![workload-resize-flowchart](../img/workload-resize-flowchart.png)
+
+### Risks and Mitigations
+
+- The main risk is the impact on the cluster when a large number of Pods adjust resources, which is entirely dependent on the user setting the maxProposed value correctly
```

Comment:
What will happen if a large number of Pods adjust resources?

Reply:
As a user, I may be concerned that if I want to make resource adjustments to the workload, but this operation will lead to pod reconstruction, the cost of this operation may be too high.

Reply:
I may be very concerned about how to achieve dynamic scaling with minimal overhead.

Reply:
@Hanmengnan After thorough analysis, I believe that it is a better choice not to separate vertical scaling as a separate operation, but to treat it as a special upgrade operation like in-place upgrades. This can solve many issues. Regarding the risk, I think as long as the vertical scaling feature of Kubernetes Pods is stable enough, there will be no uncontrollable risks in implementing it at the workload level.